### PR TITLE
Theta: The Cassandra Reaper

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
==COMMIT_MSG==
Cassandra connections are no longer removed from the pool as soon as they are no longer present in the token range; instead, they are only removed if they are absent from N consecutive token range refreshes, where N is a configurable value in `CassandraKeyValueServiceConfig`. This is expected to make the most difference when the hosts are changing frequently.
==COMMIT_MSG==

**Implementation Description (bullets)**:
* Introduced a `CassandraHostReaper` which is responsible for determining when hosts have been selected for reaping on consecutive runs, and proposes for those hosts to be deleted iff they cross the threshold configured.
* Also updated the `CassandraLogHelper` to _actually_ log the IP address and the host name.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a test for the Reaper.

**Concerns (what feedback would you like?)**:
* What default config should we set? Is 10 too high? Should we go with 1 and tune in our services?
* Can someone please confirm that the methods I'm invoking in the log helper do not perform (reverse) DNS lookups?

**Where should we start reviewing?**:
Diff.

**Priority (whenever / two weeks / yesterday)**:
ASAP

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 1]
```diff
@@ -311,27 +310,15 @@ private synchronized void refreshPool() {
...
-        serversToRemove.addAll(Sets.difference(cachedServers, desiredServers));
+        Set<InetSocketAddress> serversToAdd = Sets.difference(desiredServers, cachedServers);
+        Set<InetSocketAddress> absentServers = Sets.difference(cachedServers, desiredServers);
+        Set<InetSocketAddress> serversToRemove = reaper.computeServersToRemove(absentServers);
 
         serversToAdd.forEach(cassandra::addPool);
         serversToRemove.forEach(cassandra::removePool);
 
         if (!(serversToAdd.isEmpty() && serversToRemove.isEmpty())) { // if we made any changes
-            // Log IP addresses along with hostnames
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:325]: makes sense. I think we should pass in the _not_ removed hosts / absent hosts into the log message later for ease of reading

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraLogHelper.java [hunk 2]
```diff
@@ -68,4 +69,18 @@ private static String getUpperEndpoint(Range<LightweightOppToken> range) {
...
+    @Value.Immutable
+    interface HostAndIpAddress {
+        String host();
+
+        String ipAddress();
+
+        static HostAndIpAddress fromAddress(InetSocketAddress address) {
+            return ImmutableHostAndIpAddress.builder()
+                    .host(address.getHostString())
+                    .ipAddress(address.getAddress().getHostAddress())
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraLogHelper.java:82]: This will NPE if the address is unresolved.

**Review by jeremyk-91 — COMMENTED** [review]
Broadly makes sense!

For the reaper: I checked it's similar to https://github.com/palantir/atlasdb/pull/5862/files, didn't do much review beyond that.

For the logging: we probably want the ip address to be Optional since unresolved hostnames are a thing, and we (rightly) aren't attempting resolution in this class ourselves.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraLogHelper.java (hunk 2 above)
Comment (Jolyon-S) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraLogHelper.java:82]: oops - the first time I did this in the RC I did realise this, but obviously forgot about it now!

**Review by jeremyk-91 — APPROVED** [review]
👍

**Comment by Jolyon-S** [comment]
👍

**Pull request merged**